sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -3,16 +3,94 @@
+#    and avoid unnecessary round trips of large distance arrays with the RAM
+#    that would otherwise severely degrade the speed by making the overall
+#    processing memory-bound.
+#
+#    Finally, the routines follow a generic parallelization template to process
+#    chunks of data with OpenMP loops (via Cython prange), either on rows of X
+#    or rows of Y depending on their respective sizes.
+#
+#
+# Dispatching to specialized implementations
+# ------------------------------------------
+#
+#    Dispatchers are meant to be used in the Python code. Under the hood, a
+#    dispatcher must only define the logic to choose at runtime to the correct
+#    dtype-specialized :class:`PairwiseDistancesReduction` implementation based
+#    on the dtype of X and of Y.
+#
+#
+# High-level diagram
+# ------------------
+#
+#   Legend:
+#
+#      A ---⊳ B: A inherits from B
+#      A ---x B: A dispatches on B
+#      A ---* B: A composes B
+#
+#
+#                           (base dispatcher)
+#                       PairwiseDistancesReduction
+#                                   ∆
+#                                   |
+#                                   |
+#                 +-----------------+-----------------+
+#                 |                                   |
+#           (dispatcher)                        (dispatcher)
+#     PairwiseDistancesArgKmin           PairwiseDistancesRadiusNeighbors
+#           |                                              |
+#           |                                              |
+#           |                                              |
+#           |                  (64bit implem.)             |
+#           |          PairwiseDistancesReduction64        |
+#           |                       ∆                      |
+#           |                       |                      |
+#           |                       |                      |
+#           |     +-----------------+-----------------+    |
+#           |     |                                   |    |
+#           |     |                                   |    |
+#           x     |                                   |    x
+#    PairwiseDistancesArgKmin64       PairwiseDistancesRadiusNeighbors64
+#           |     ∆                                   ∆    |
+#           |     |                                   |    |
+#           x     |                                   |    |
+# FastEuclideanPairwiseDistancesArgKmin64             |    |
+#                 *                                   |    |
+#                 |                                   |    x
+#                 |            FastEuclideanPairwiseDistancesRadiusNeighbors64
+#                 |                                   *
+#                 |                                   |
+#                 +-----------------+-----------------+
+#                                   |
+#                            GEMMTermComputer64
```

Comment:
`GEMMTermComputer64` feels more like an implementation detail and does not need to be a part of this diagram?

Reply:
That's right. I guess we can come up with another diagram for implementations which would explain technical details.

Reply:
If one of the goals is https://github.com/scikit-learn/scikit-learn/pull/23515#discussion_r894919696, then "circle" in this diagram can be confusing. One does not need to know about the implementation detail of `GEMMTermComputer64` to learn about the dispatching mechanism.

I prefer a sentence or two to say `GEMMTermComputer64` is used by `FastEuclideanPairwiseDistancesRadiusNeighbors64` and `FastEuclideanPairwiseDistancesArgKmin64`.